# Simulating Monty Hall's classic problem!

In [1]:
from game import Game
import random

In [2]:
def play_game(switch: bool = False) -> bool:
    """
    Play a game, randomly select a door, default to not switch door. Return Win-ness
    """
    game = Game()
    game.player_select(random.randint(0, 2))
    if switch:
        game.player_switch()

    if game.players_door.prized:
        return True
    return False

False

In [3]:
def simulate_samples(size: int, switch: bool = False):
    wins = 0
    for _ in range(size):
        if play_game(switch=switch):
            wins += 1
    
    print(f"After playing {size} games, Won {wins}. Probability: {wins/(size+1)}")

#### Simulate different ranges of sample sizes, never switch doors:

In [4]:
simulate_samples(1_000)
simulate_samples(10_000)
simulate_samples(100_000)

After playing 1000 games, Won 326. Probability: 0.3256743256743257
After playing 10000 games, Won 3418. Probability: 0.3417658234176582
After playing 100000 games, Won 33436. Probability: 0.33435665643343565


#### Simulate different ranges of sample sizes, always switch doors:

In [5]:
simulate_samples(1_000, switch=True)
simulate_samples(10_000, switch=True)
simulate_samples(100_000, switch=True)

After playing 1000 games, Won 671. Probability: 0.6703296703296703
After playing 10000 games, Won 6718. Probability: 0.6717328267173283
After playing 100000 games, Won 66785. Probability: 0.6678433215667844


## Conclusion:
Always switch doors. From the simulation above, not switchind door yealds approximately 33% winrate while switching doors yealds approximately 66% winrate. vos Savant was right =))